_posts/2022-03-15-turn-based-offline-rl.md

```diff
@@ -0,0 +1,344 @@
+            trajectories_from_new_policy = deploy_and_sample(current_policy, num_of_trajectories_per_turn)
+            temp_dataset.extend(trajectories_from_new_policy)
+            policy, performances = run_offline_rl_algorithm(temp_dataset, num_iters)
+            runs.append((policy, performances, trajectories_from_new_policy))
+
+        best_policy, best_trajectories = find_best_run(runs)
+        current_policy = best_policy
+        offline_dataset.extend(best_trajectories)
+```
+
+Let’s explain each step involved in the algorithm.  First, let’s define what are the main parameters expected by
+the algorithm:
+
+* `init_policy` — it’s the starting policy, most likely the random policy.
+* `num_turns` — this is simply the total number of turns for which you will run the algorithm.
+* `num_seeds` — if you work in RL you will be familiar with this argument: RL algorithms
+(and especially Offline RL ones) present large variability in the results due to their stochastic nature.
+That’s why instead of having one single run of the Offline RL algorithm,
+we will have several of them. For each of them, we will produce the best policy and the best
+“new set of trajectories” to be used later in the algorithm (more on this step in the following).
+* `num_iters` — this is simply the number of iterations we will run our Offline RL algorithm.
+* `dataset_size` — as a design choice, we assume that the final dataset size has been fixed beforehand,
+as we do with the number of turns. However, both of these two conditions could be relaxed and one could run
+the algorithm as many turns as needed, getting a final offline dataset with an undefined size.
+However, please remember that in the real scenario you will probably not have the privilege of
+accessing the environment so often! You must do your best with a reasonable number of turns!
+
+Now, following the logic of the pseudo-code, let’s describe the algorithm:
+
+* Initially, we don’t have any transitions to train our Offline RL algorithm, so we initialize our `offline_dataset`
+as an empty list.
+* We also initialize `current_policy` with `init_policy`, which most likely will be the random policy
+(an agent that has previously interacted with the environment taking actions uniformly at random).
+* Now, for each turn we will run `num_seeds` times the following procedure:
+    * We create a copy of `offline_dataset` (`temp_dataset`) to train the current agent with the dataset collected
+  so far.
+    * We deploy the agent to the environment, in order to generate a new set of transitions using the current policy
+  (`trajectories_from_new_policy`).
+    * We extend the temporary dataset by `trajectories_from_new_policy` and train an agent with it,
+  using the preferred Offline RL algorithm and getting its corresponding `policy` and `performances`.
+    * We append the results to `runs` list.
+    * Once we have collected all the results, we will pick the best policy and best-generated trajectories
```

Comment:
"will" zbędne

Reply:
I've changed it and also "normalized" the entire paragraph to using the "present" and not the "future"